feat(replication): Use run_queue to avoid CouchDB overload

H--o-l opened the pull request:
feat(replication): Use run_queue to avoid CouchDB overload

A few commits have try to avoid or recover from error that happen when
compute limitations are reached:
[`feat(replication): Handle recoverable problems`]
[`fix(replication): timeout`]
[`feat(replication): Handle recoverable problems`]
[`feat(replication): Use a dynamic replication speed`]
[`feat(CouchDB): Try to avoid requests timeout on replication`]
[`feat(CouchDB): Limit number of replication worker processes`]

But all of these commit allow `coucharchive` to run while machines themselves
(source or target) may be at a breaking point.
Depending what process inside CouchDB start to fail, CouchDB could recover or
it could go wild.

A solution I propose, to control the load on machines, source and target, is
to use [CouchDB `/_node/{node-name}/_system` API], with `_local` as `node-name`.
Especially, the `run_queue` info returned by the API seems a good representation
of compute limitations.
[Citing Adam Kocoloski]:
```
The “run_queue” is the number of Erlang processes in a runnable state that are
not currently executing on a scheduler. When that value is greater than zero it
means the node is hitting some compute limitations. Seeing a small positive
value from time to time is no problem.
```
The CLI option `--no-run-queue-control` can be set to true to restore the
previous behavior, if CouchDB errors are not a big deal in some use case.

A timeout parameter of 60 secondes is used when calling
[CouchDB `/_node/{node-name}/_system` API] in case this API keep the socket open
and do not answer. It happen to myself only once over hundred of tests, but it
can be critical to avoid overload.

[Citing Adam Kocoloski]: http://mail-archives.apache.org/mod_mbox/couchdb-user/201906.mbox/<90FEEE0B-B2E5-4B72-A196-4A35E26C467B%40apache.org>
[CouchDB `/_node/{node-name}/_system` API]: https://docs.couchdb.org/en/stable/api/server/common.html#node-node-name-system
[`feat(replication): Handle recoverable problems`]: https://github.com/adrienverge/coucharchive/commit/d1c0941955bf64a2e7b6714c07ea516dca50c5e7
[`fix(replication): timeout`]: https://github.com/adrienverge/coucharchive/commit/6c0863246f4da014bd2b93e6c0cd80b9e8f747cc
[`feat(replication): Handle recoverable problems`]: https://github.com/adrienverge/coucharchive/commit/4a326fe9bcccd326434e8cc98d8f0539f817bb5f
[`feat(replication): Use a dynamic replication speed`]: https://github.com/adrienverge/coucharchive/commit/1ea19faf72dd9442ecd1eb7a89e954b1c4c687a3
[`feat(CouchDB): Try to avoid requests timeout on replication`]: https://github.com/adrienverge/coucharchive/commit/a8d764e768a4529045c19b539ea2ad8ad8426555
[`feat(CouchDB): Limit number of replication worker processes`]: https://github.com/adrienverge/coucharchive/commit/b8125a929879d3180667bfb3804509c1971201ae

H--o-l committed:
feat(replication): Use run_queue to avoid CouchDB overload

A few commits have try to avoid or recover from error that happen when
compute limitations are reached:
[`feat(replication): Handle recoverable problems`]
[`fix(replication): timeout`]
[`feat(replication): Handle recoverable problems`]
[`feat(replication): Use a dynamic replication speed`]
[`feat(CouchDB): Try to avoid requests timeout on replication`]
[`feat(CouchDB): Limit number of replication worker processes`]

But all of these commit allow `coucharchive` to run while machines themselves
(source or target) may be at a breaking point.
Depending what process inside CouchDB start to fail, CouchDB could recover or
it could go wild.

A solution I propose, to control the load on machines, source and target, is
to use [CouchDB `/_node/{node-name}/_system` API], with `_local` as `node-name`.
Especially, the `run_queue` info returned by the API seems a good representation
of compute limitations.
[Citing Adam Kocoloski]:
```
The “run_queue” is the number of Erlang processes in a runnable state that are
not currently executing on a scheduler. When that value is greater than zero it
means the node is hitting some compute limitations. Seeing a small positive
value from time to time is no problem.
```
The CLI option `--no-run-queue-control` can be set to true to restore the
previous behavior, if CouchDB errors are not a big deal in some use case.

A timeout parameter of 60 secondes is used when calling
[CouchDB `/_node/{node-name}/_system` API] in case this API keep the socket open
and do not answer. It happen to myself only once over hundred of tests, but it
can be critical to avoid overload.

[Citing Adam Kocoloski]: http://mail-archives.apache.org/mod_mbox/couchdb-user/201906.mbox/<90FEEE0B-B2E5-4B72-A196-4A35E26C467B%40apache.org>
[CouchDB `/_node/{node-name}/_system` API]: https://docs.couchdb.org/en/stable/api/server/common.html#node-node-name-system
[`feat(replication): Handle recoverable problems`]: https://github.com/adrienverge/coucharchive/commit/d1c0941955bf64a2e7b6714c07ea516dca50c5e7
[`fix(replication): timeout`]: https://github.com/adrienverge/coucharchive/commit/6c0863246f4da014bd2b93e6c0cd80b9e8f747cc
[`feat(replication): Handle recoverable problems`]: https://github.com/adrienverge/coucharchive/commit/4a326fe9bcccd326434e8cc98d8f0539f817bb5f
[`feat(replication): Use a dynamic replication speed`]: https://github.com/adrienverge/coucharchive/commit/1ea19faf72dd9442ecd1eb7a89e954b1c4c687a3
[`feat(CouchDB): Try to avoid requests timeout on replication`]: https://github.com/adrienverge/coucharchive/commit/a8d764e768a4529045c19b539ea2ad8ad8426555
[`feat(CouchDB): Limit number of replication worker processes`]: https://github.com/adrienverge/coucharchive/commit/b8125a929879d3180667bfb3804509c1971201ae

bravier commented on coucharchive:
```diff
@@ -76,6 +77,31 @@ def _check_couchdb_connection(url):
+    response = urllib.request.urlopen(req)
+    data = json.loads(response.read().decode('utf-8'))
+
+    return data
```
It can also be written as:
```python
return json.loads(response.read().decode('utf-8'))
```
to avoid declaring `data` variable.

H--o-l replied:
Oh yeah. Thanks @bravier for the review, corrections are pushed!

bravier commented on coucharchive:
```diff
@@ -261,8 +287,25 @@ def make_throttler_faster():
+        time.sleep(0.5)  # arbitrary
+        should_slow = False
+        for url in urls:
+            couch_queue = _get_couch_system_stats(url)['run_queue']
```
Will `run_queue` always be present in the response?

H--o-l replied:
I think that it will.

adrienverge replied:
If I understand correctly, we will do a request to each server every 0.5 second. I seems a lot, no? Can this participate to the server overload?

Is it an option to call this route only every 5 seconds, but use something like `make_throttler_slower(10×)`?

H--o-l replied:
> If I understand correctly, we will do a request to each server every 0.5 second.

Indeed.

> I seems a lot, no? Can this participate to the server overload?

I was wondering too, but replications them-self seems way worse.

Reason I'm saying that is because CouchDB seems to be really quick to answer to that API event when there is huge load (otherwise it would be impractical for this use case).
I think the data are pre-calculated and just dumped to the API when called.
So the added charged are only on the TCP stack that have a bit more work to do, but all in all I think it's reasonable.

And benefits are:
- no code modification of the throttler
- quicker and smoother control loop

*10 each 5 second seems low. It means we will need 3 consecutive calls (15 seconds) to reach a 60 seconds throttler:
```
0s --------> 1s -----------> 20s ----------> 400s (top at 60s)
     +1           x2x10
```
Today we reach a 60 seconds throttler in 7 calls (3.5 seconds):
```
0s ---> 1s ---> 2s ---> 4s ---> 8s ---> 16s ---> 32s ---> 64s  (top at 60s)
    +1     x2      x2     ...
```

What do you think of :
- call the API each two seconds
- modification of the throller value direcly:
  - when raising up, If equals `0` then make it `30`, otherwise multiple by 20
  - also lower it quicker (it's a default with the current PR code it lower slowly) by dividing by 20 or something like that.
  - maybe another smarter solution, modification of the throttler value directly but using `x = run_queue * 5` for exemple

adrienverge replied:
Very interesting!

> Today we reach a 60 seconds throttler in 7 calls (3.5 seconds):

I don't get that: today, we don't call `make_throttler_slower()` every *n* seconds: it's called only when there is an error, isn't? So it doesn't take "3.5 seconds", in average it should be more like ~ 30 seconds, no?

Anyway:
- +1 to call the API every two seconds
- > If equals 0 then make it 30, otherwise multiple by 20

  What?!? :smile: It means if  `make_throttler_slower()` is called twice, it will delay all replications by `30 × 20 = 600` seconds?

- the other solution (using the run queue value) is smart indeed, it can be a good one!

H--o-l replied:
You may have been right, I do not remember sorry.
The code on which this PR is changed a bit, so I will resolve the conversation.
If the issue is still in the new code don't hesitate to open a new one!

bravier commented on coucharchive:
```diff
@@ -522,6 +581,11 @@ def main():
+    sub['create'].add_argument(
+        '--no-queue-control', dest='no_queue_control',
+        action='store_true', default=False,
+        help='de-activate throttle request if origin couchdb `run_queue` '
```
`deactivate`: no need for the dash ;)

H--o-l replied:
Ok!

bravier reviewed (commented):
Hi Hoël!
Your PR seems to be a very good idea!
I've just some questions/comments on it.

adrienverge commented on coucharchive:
```diff
@@ -76,6 +77,30 @@ def _check_couchdb_connection(url):
+        'Authorization': 'Basic %s' % auth})
+    response = urllib.request.urlopen(req)
+
+    return json.loads(response.read().decode('utf-8'))
```
To prevent code duplication, I propose to reuse `_couchdb_request()` from https://github.com/adrienverge/coucharchive/pull/20, would that be OK for you?

H--o-l replied:
It have been perfect thanks :-)

adrienverge commented on coucharchive:
```diff
@@ -32,6 +32,7 @@ import couchdb
 
 
 NUMBER_OF_WORKERS = 64
+QUEUE_LENGTH_THRESHOLD = 7
```
To make this easy for new developers, I propose:
```diff
-QUEUE_LENGTH_THRESHOLD
+COUCHDB_MAX_RUN_QUEUE
```
What do you think?

H--o-l replied:
It's a good idea!

adrienverge commented on coucharchive:
```diff
@@ -274,9 +316,19 @@ def replicate_couchdb_server(source_url, target_url,
     todos = [(source_url, target_url, db, reuse_db_if_exist) for db in all_dbs]
 
+    if queue_control:
+        queue_control_thread = \
```
If it's not a thread, I'd rather not call it a thread, it could be confusing.
- Let's make it a thread? (best option, I think)
- Or rename it to `queue_control_process`?

H--o-l replied:
> Let's make it a thread? (best option, I think)

Would I be able to access the current throttler implementation from a thread?

adrienverge replied:
Sure! A thread is "closer" than an another process, i.e. a thread has access to more data from the original process, than another process.

H--o-l replied:
Thanks for the explanation!

adrienverge commented on coucharchive:
```diff
@@ -274,9 +316,19 @@ def replicate_couchdb_server(source_url, target_url,
+            if queue_control and not queue_control_thread.is_alive():
+                pool.close()
+                pool.terminate()
+                exit(1)
```
What is `exit()`? Can you explain this `for` loop please? I'm not sure to understand it properly...

H--o-l replied:
OK, that part was really hard to do.
Idea is that I want a process that run in parrallel of the other, and that make the program stop if it failed for a reason.
But the way the other process are called do not let a lot of choice.
So here what happen is that each time a replication process finish, the queue control process is checked (with `.is_alive()`).
If it's not alive it close the replication pool of worker, and terminate the program with an error.
Exceptions display of all workers appear correctly from what I have seen.
I'm would be happy to try another implementation if you have an idea!

H--o-l replied:
(PS: I will change `exit(1)` in `sys.exit(1)`)

adrienverge commented on coucharchive:
```diff
@@ -522,6 +580,11 @@ def main():
+    sub['create'].add_argument(
+        '--no-queue-control', dest='no_queue_control',
+        action='store_true', default=False,
+        help='deactivate throttle request if origin couchdb `run_queue` '
```
Sorry for the double modification (following @bravier's remark), but *deactivate → disable* would be more standard I think!

Other proposals:
- throttle request → throttling of requests
- ~origin~  *(I think now it's both source and target)*
- couchdb → CouchDB
- `run_queue` system info is too high → run queue is too long

H--o-l replied:
Thanks!
It give:
```py
        help='disable throttling of requests if CouchDB `run_queue` is too '
             'high')
```

adrienverge commented on coucharchive:
```diff
@@ -534,6 +597,11 @@ def main():
         action='store_true', default=False,
         help='continue restoration even if database exists on target')
+    sub['restore'].add_argument(
+        '--no-queue-control', dest='no_queue_control',
```
Sometimes we use "queue", sometimes "run_queue". One day, we could implement our own queue inside coucharchive.

To avoid problems in the future, I propose to use "run_queue" and `--no-run-queue-control` everywhere. What do you think?

H--o-l replied:
First time I was thinking of a parameters called `hardware-monitoring` or something like that, because if we realize that their is a better parameter to check than `run_queue` we would have to rename it all.
Anyway,  `--no-run-queue-control` is better than `--no-queue-control`, I will rename it, thanks!

adrienverge commented on coucharchive:
```diff
@@ -498,10 +554,12 @@ def load(filename):
     def callback(local_couch_server_url):
         replicate_couchdb_server(local_couch_server_url, target,
                                  reuse_db_if_exist=reuse_db_if_exist,
+                                 queue_control=queue_control,
```
This one is not very important, it's more a cherry on the cake... but I propose to use `no_run_queue_control` or `has_run_queue_control` in Python arguments. Otherwise, when I read this, it looks like you're passing an object.

H--o-l replied:
Clever, I will use `has_run_queue_control` if it's OK.

adrienverge reviewed (commented):
Wow, it really looks promising! Can't wait to test it on a big cluster.

Please see my few comments!

H--o-l commented:
Hey @adrienverge @bravier !
Following your reviews here + review here https://github.com/adrienverge/coucharchive/commit/56511419ef56008d4ee2b728e0a147d97735b8da, and a rebase on the last master version, I pushed a new version.

It's not tested, I will test it right now and will keep you informed.

adrienverge commented on coucharchive:
```diff
@@ -255,6 +257,35 @@ class ReplicationControl(object):
 
+        self.has_run_queue_control = has_run_queue_control
+        if self.has_run_queue_control:
+            def run_queue_control():
```
Suggestion for readability and easier indentation: make it a class function! :slightly_smiling_face: 
```python
class ReplicationControl:
    def _run_queue_control_thread(self):
```

H--o-l replied:
:+1:

adrienverge commented on coucharchive:
```diff
@@ -255,6 +257,35 @@ class ReplicationControl(object):
+            def run_queue_control():
+                t = threading.currentThread()
+                while getattr(t, "do_run", True):
+                    time.sleep(2)  # arbitrary
```
This line would be better at the end of the loop, otherwise, when we decide to stop, the thread will still be "survive-waiting" for 2 seconds.

EDIT: Actually, it will "survive-wait" in any case... but still, I find it more logical to do the request first, rather than doing an extra request at the end, just before exiting.

H--o-l replied:
OK.

adrienverge commented on coucharchive:
```diff
@@ -255,6 +257,35 @@ class ReplicationControl(object):
+        if self.has_run_queue_control:
+            def run_queue_control():
+                t = threading.currentThread()
+                while getattr(t, "do_run", True):
```
Suggestion for understanding (take it or leave it):
```python
                t = threading.currentThread()
                t.do_run = True
                while t.do_run:
```

H--o-l replied:
I did not understood, do you mean something like that?
```py
                t = threading.currentThread()
                t.do_run = getattr(t, "do_run", True)
                while t.do_run:

adrienverge replied:
No, I meant the code I posted, with `t.do_run = True` set once at the beginning, then loop until another thread does `t.do_run = False`. Maybe I missed something? Otherwise, I find my proposal more readable!

H--o-l replied:
Ah yes, it works great indeed, thanks!

adrienverge commented on coucharchive:
```diff
@@ -255,6 +257,35 @@ class ReplicationControl(object):
+                    logging.debug('CouchDB run_queue: %d' % run_queue)
+
+            self.thread = \
+                threading.Thread(target=run_queue_control)
```
This could go in one line :wink: 

Also, maybe `self.run_queue_thread` would be more directly understandable, inside `self.close()`?

H--o-l replied:
I changed:
- `self.thread` to `self.run_queue_control_thread`
- `def run_queue_control():` to `self._run_queue_control_function()`

What do you think of it?

adrienverge replied:
:+1: 

(no need for the `_function()` suffix in my opinion, but OK with it too)

adrienverge commented on coucharchive:
```diff
@@ -661,6 +697,11 @@ def main():
+        '--no-run-queue-control', dest='no_run_queue_control',
+        action='store_true', default=False,
+        help='disable throttling of requests if CouchDB `run_queue` is too '
+             'high')
```
Since #22, there is no real "throttling of requests", so I propose something like "don't read CouchDB run_queue length to adapt the number of concurrent replications", what do you think?

H--o-l replied:
Good catch! Your proposal looks ok to me, I will use it, thanks!

adrienverge commented on coucharchive:
```diff
@@ -342,13 +374,14 @@ def replicate_couchdb_server(source_url, target_url,
     error = None
 
-    while len(dbs) and not error:
+    while len(dbs) and not error and control.is_alive():
```
If I understand correctly:
1. If the run_queue thread dies (e.g. random network problem), it will stop coucharchive.
2. If the run_queue thread dies, it won't raise any error nor bad exit code: the user won't know that replication was stopped in the middle.

H--o-l replied:
The error trace is displayed by the thread itself.
So it doesn't return an error like replications process, but it's displayed in `stderr` like other traces of `coucharchive`.
For example, raising a fake `toto` exception in the thread give:
```python trace
INFO: blabla: done
INFO: Oo: done
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib64/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "./coucharchive", line 273, in _run_queue_control_function
    raise Exception('toto')
Exception: toto

INFO: Terminating local CouchDB instance
INFO: Creating backup archive at toto.tar.gz
```

adrienverge replied:
OK, so we have a problem, no? If the backup file is created and no error code returned when the control thread dies and the replication loop stopped, the user will think it's OK while it's not? (by the way, the traceback will be way up in logs, if you have 256 workers)

H--o-l replied:
Hey Adrien!
You are right I haven't been carrefull with error return code of the program, which is wrong in this case.
I have too solution to propose you:
1. https://github.com/adrienverge/coucharchive/commit/ca1b93a8c5f4d424bd2539c29506ff5ac39d5707
2. https://github.com/adrienverge/coucharchive/commit/42e9072516378e2be765420060f7bb6c69815bf5

I let you tell me if you like one.
On my part, over the two, I prefer the second, because it re-use the `getattr(t, "do_run", True)` to store the exception, and thus handle the issue you mentioned in bracket (`by the way, the traceback will be way up in logs, if you have 256 workers`)

H--o-l replied:
Subject discussed in private messages, solution 1. chose!
```diff
--- a/coucharchive
+++ b/coucharchive
@@ -396,7 +396,7 @@ def replicate_couchdb_server(source_url, target_url,
                                  target_url, reuse_db_if_exist))
     error = None
 
-    while len(dbs) and not error and control.is_alive():
+    while len(dbs) and not error:
         ideal = control.ideal_number_of_replications()
         while len(dbs) and control.running_replications.value < ideal:
             in_queue.put(dbs.pop(0))
@@ -415,6 +415,10 @@ def replicate_couchdb_server(source_url, target_url,
 
         time.sleep(1)
 
+        if not control.is_alive():
+            error = Exception('Control died.')
+            break
+
         while True:
             try:
                 result = out_queue.get_nowait()

adrienverge reviewed (commented):
It looks promising! I have a few remarks, and I'm looking forward to see results!


PS: `feat(replication): Use CouchDB run_queue to avoid CouchDB overload` would be :ok_hand:

H--o-l commented:
Hum, with the current implementation CPU utilization oscillate heavily on my machines:
![toto](https://user-images.githubusercontent.com/10905922/68749415-7a4e3800-05fe-11ea-8ab3-440654e45f2b.png)
And It even end-up crashing :-/
(it's a backup from a source bleu machine and a target yellow machine).

Maybe the crash is due to the fact that the algorythm do not react quick enought?
In log I saw that `run_queue` sometime goes high, and takes time do go down.
log here: 
[screenlog.zip](https://github.com/adrienverge/coucharchive/files/3840326/screenlog.zip)

H--o-l commented:
Hey!

I change some parameters, see the diff below, and I have a much smoother curve:
![toto](https://user-images.githubusercontent.com/10905922/68781822-a1c1f680-0638-11ea-971d-8b9ac2835af9.png)
(Still backup bleu the source, yellow the target)
I think the CPU utilization at start is too high compare to what this PR want to achieve.

From what I see in the log, worker count decrease correctly when `run_queue > 7` but as we start at a 128 worker count, these 128 are still running and trying to access CouchDB while it is having a hard time.

I think it would to better to top the number of worker at 64, so we quickly stop requesting the CouchDB in difficulty.

In the same time I will retry parameter of my previous post, just to be sure.

the diff:
```diff
--- a/coucharchive
+++ b/coucharchive
@@ -296,7 +296,7 @@ class ReplicationControl(object):
                                   self.running_replications.value))
 
     def _drop_old_events(self):
-        five_min_ago = time.time() - 5 * 60
+        five_min_ago = time.time() - 20
         self._last_successes[:] = [e for e in self._last_successes
                                    if e[0] > five_min_ago]
         self._last_errors[:] = [e for e in self._last_errors
@@ -321,7 +321,7 @@ class ReplicationControl(object):
             self.ideal_speed = databases_left / time_left
             self.current_avge_speed = (
                 len(self._last_successes) /
-                min(time.time() - self._start_time, 5 * 60))
+                min(time.time() - self._start_time, 20))
             current_avge_replications = (
                 sum(e[1] for e in self._last_successes) /
                 len(self._last_successes))
@@ -342,7 +342,7 @@ class ReplicationControl(object):
         # ... but if there were errors, take 0.9 × the lowest error value
         if self._last_errors:
             min_number_with_error = min(e[1] for e in self._last_errors)
-            ideal_number = int(0.9 * min_number_with_error)
+            ideal_number = int(0.3 * min_number_with_error)
 
         # ... and stay within [1, MAX_NUMBER_OF_WORKERS]
         ideal_number = max(1, min(MAX_NUMBER_OF_WORKERS, ideal_number))
```

The log: 
[screenlog.zip](https://github.com/adrienverge/coucharchive/files/3842260/screenlog.zip)

H--o-l commented:
I confirm my results of yesterday (the two last comments) with another test this night:
![toto](https://user-images.githubusercontent.com/10905922/68837550-176fa600-06bd-11ea-8b89-98bd41aef83b.png)

So I think that for the `run_queue` parameter to be used, some values of the existing algorithm have to be adjusted.
@adrienverge what do you think is the best way of doing it?
Should the values be adjusted only when `run_queue` is used?

H--o-l commented:
While doing a backup test I had a crash:
```
DEBUG: Currently running 95 replication workers
DEBUG: Ideal speed = fastest; current average speed = 6.8 rep/s
DEBUG: There were 1 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128)]
DEBUG: Currently running 85 replication workers
DEBUG: Ideal speed = fastest; current average speed = 6.8 rep/s
DEBUG: There were 1 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128)]
DEBUG: Currently running 79 replication workers
DEBUG: Ideal speed = fastest; current average speed = 6.8 rep/s
DEBUG: There were 1 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128)]
DEBUG: Currently running 76 replication workers
DEBUG: Ideal speed = fastest; current average speed = 6.5 rep/s
DEBUG: There were 1 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128)]
DEBUG: CouchDB run_queue: 63
DEBUG: Currently running 73 replication workers
DEBUG: Ideal speed = fastest; current average speed = 6.5 rep/s
DEBUG: There were 2 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76)]
DEBUG: Currently running 72 replication workers
DEBUG: Ideal speed = fastest; current average speed = 6.2 rep/s
DEBUG: There were 2 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76)]
DEBUG: Currently running 71 replication workers
DEBUG: Ideal speed = fastest; current average speed = 5.8 rep/s
DEBUG: There were 2 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76)]
DEBUG: Currently running 71 replication workers
DEBUG: Ideal speed = fastest; current average speed = 5.5 rep/s
DEBUG: There were 2 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76)]
DEBUG: CouchDB run_queue: 15
DEBUG: Currently running 70 replication workers
DEBUG: Ideal speed = fastest; current average speed = 4.8 rep/s
DEBUG: There were 3 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76), (1573737169, 71)]
DEBUG: Currently running 70 replication workers
DEBUG: Ideal speed = fastest; current average speed = 4.3 rep/s
DEBUG: There were 3 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76), (1573737169, 71)]
DEBUG: Currently running 70 replication workers
DEBUG: Ideal speed = fastest; current average speed = 4.2 rep/s
DEBUG: There were 3 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76), (1573737169, 71)]
DEBUG: CouchDB run_queue: 3
DEBUG: Currently running 70 replication workers
DEBUG: Ideal speed = fastest; current average speed = 3.9 rep/s
DEBUG: There were 3 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76), (1573737169, 71)]
DEBUG: Currently running 68 replication workers
DEBUG: Ideal speed = fastest; current average speed = 3.5 rep/s
DEBUG: There were 3 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76), (1573737169, 71)]
DEBUG: Currently running 68 replication workers
DEBUG: Ideal speed = fastest; current average speed = 3.1 rep/s
DEBUG: There were 3 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76), (1573737169, 71)]
DEBUG: CouchDB run_queue: 23
DEBUG: Currently running 67 replication workers
DEBUG: Ideal speed = fastest; current average speed = 3.0 rep/s
DEBUG: There were 4 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737156, 128), (1573737164, 76), (1573737169, 71), (1573737175, 68)]
DEBUG: Currently running 67 replication workers
DEBUG: Ideal speed = fastest; current average speed = 3.0 rep/s
DEBUG: There were 3 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737164, 76), (1573737169, 71), (1573737175, 68)]
DEBUG: Currently running 65 replication workers
DEBUG: Ideal speed = fastest; current average speed = 2.8 rep/s
DEBUG: There were 3 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737164, 76), (1573737169, 71), (1573737175, 68)]
DEBUG: Currently running 65 replication workers
DEBUG: Ideal speed = fastest; current average speed = 2.4 rep/s
DEBUG: There were 3 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737164, 76), (1573737169, 71), (1573737175, 68)]
DEBUG: CouchDB run_queue: 10
DEBUG: Currently running 65 replication workers
DEBUG: Ideal speed = fastest; current average speed = 1.9 rep/s
DEBUG: There were 4 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737164, 76), (1573737169, 71), (1573737175, 68), (1573737179, 65)]
DEBUG: Currently running 65 replication workers
DEBUG: Ideal speed = fastest; current average speed = 1.6 rep/s
DEBUG: There were 4 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737164, 76), (1573737169, 71), (1573737175, 68), (1573737179, 65)]
DEBUG: CouchDB run_queue: 20
DEBUG: Currently running 64 replication workers
DEBUG: Ideal speed = fastest; current average speed = 1.2 rep/s
DEBUG: There were 5 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737164, 76), (1573737169, 71), (1573737175, 68), (1573737179, 65), (1573737181, 65)]
DEBUG: Currently running 64 replication workers
DEBUG: Ideal speed = fastest; current average speed = 0.8 rep/s
DEBUG: There were 5 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737164, 76), (1573737169, 71), (1573737175, 68), (1573737179, 65), (1573737181, 65)]
DEBUG: Currently running 64 replication workers
DEBUG: Ideal speed = fastest; current average speed = 0.7 rep/s
DEBUG: There were 5 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737164, 76), (1573737169, 71), (1573737175, 68), (1573737179, 65), (1573737181, 65)]
DEBUG: CouchDB run_queue: 88
DEBUG: Currently running 64 replication workers
DEBUG: Ideal speed = fastest; current average speed = 0.5 rep/s
DEBUG: There were 5 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737169, 71), (1573737175, 68), (1573737179, 65), (1573737181, 65), (1573737184, 64)]
DEBUG: Currently running 64 replication workers
DEBUG: Ideal speed = fastest; current average speed = 0.4 rep/s
DEBUG: There were 5 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737169, 71), (1573737175, 68), (1573737179, 65), (1573737181, 65), (1573737184, 64)]
DEBUG: Currently running 64 replication workers
DEBUG: Ideal speed = fastest; current average speed = 0.4 rep/s
DEBUG: There were 5 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737169, 71), (1573737175, 68), (1573737179, 65), (1573737181, 65), (1573737184, 64)]
DEBUG: CouchDB run_queue: 51
DEBUG: Currently running 64 replication workers
DEBUG: Ideal speed = fastest; current average speed = 0.3 rep/s
DEBUG: There were 6 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737169, 71), (1573737175, 68), (1573737179, 65), (1573737181, 65), (1573737184, 64), (1573737187, 64)]
DEBUG: Currently running 63 replication workers
DEBUG: Ideal speed = fastest; current average speed = 0.4 rep/s
DEBUG: There were 6 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737169, 71), (1573737175, 68), (1573737179, 65), (1573737181, 65), (1573737184, 64), (1573737187, 64)]
DEBUG: CouchDB run_queue: 43
DEBUG: Currently running 63 replication workers
DEBUG: Ideal speed = fastest; current average speed = 0.3 rep/s
DEBUG: There were 6 CouchDB errors encountered in the last 5 minutes
DEBUG: Last errors: [(1573737175, 68), (1573737179, 65), (1573737181, 65), (1573737184, 64), (1573737187, 64), (1573737189, 63)]
Process ForkPoolWorker-94:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 103, in worker
    initializer(*initargs)
  File "./coucharchive", line 451, in replicate_databases
    reuse_db_if_exist, item)
  File "./coucharchive", line 474, in replicate_one_database
    server.replicate(source_url + '/' + db, target_url + '/' + db)
  File "/usr/local/lib/python3.6/site-packages/couchdb/client.py", line 242, in replicate
    status, headers, data = self.resource.post_json('_replicate', data)
  File "/usr/local/lib/python3.6/site-packages/couchdb/http.py", line 574, in post_json
    **params)
  File "/usr/local/lib/python3.6/site-packages/couchdb/http.py", line 596, in _request_json
    headers=headers, **params)
  File "/usr/local/lib/python3.6/site-packages/couchdb/http.py", line 592, in _request
    credentials=self.credentials)
  File "/usr/local/lib/python3.6/site-packages/couchdb/http.py", line 429, in request
    raise ServerError((status, error))
couchdb.http.ServerError: (500, ('replication_auth_error', '{session_request_failed, ... req_timedout}'))
```

In this test I had `MAX_NUMBER_OF_WORKERS = 128`, and you can see `run_queue` is staying up for a while before the crash, and that the number of real worker running decrease (too) slowly.
I definitively think that `128` parallel workers is too high to allow this PR to regulate the charge.
I will continue testing with `MAX_NUMBER_OF_WORKERS = 64`

H--o-l commented:
Hey @adrienverge !
So for a bit of history inside this PR, the issue I describe in my last four comments were in fact due (and fixed) to PR #25.
Thus this PR is now rebased on top of #25 so I can run it.

I also integrate your last review inside the commit and pushed it, + two small mistakes (I didn't close the thread correctcly at the end, and the `load` feature was crashing on missing `--no-run-queue-control` parameter).
I'm sorry that I didn't save the diff before/after, so you may have to start a new review from scratch.

Version pushed is the version I'm currently using for my fake live test!

H--o-l commented:
Hey Adrien!

This PR have been rebased over Coucharchive v2.2.2.
Beside that, following a crash I had this night on my cluster when creating a backup, I add a small message in the commit message and a small bit of code:
```
A timeout parameter of 60 secondes is used when calling
CouchDB /_node/{node-name}/_system API in case this API keep the socket open
and do not answer. It happen to myself only once over hundred of tests, but it
can be critical to avoid overload.
```
```diff
--- a/coucharchive
+++ b/coucharchive
@@ -65,7 +65,7 @@ def _canonical_couchdb_url(url):
     return auth_url
 
 
-def _couchdb_request(url):
+def _couchdb_request(url, timeout=None):
     parts = list(urlsplit(url))
     credentials, server = parts[1].rsplit('@', 1)
     username, password = credentials.split(':', 1)
@@ -75,7 +75,10 @@ def _couchdb_request(url):
     url = urlunsplit(parts)
     req = urllib.request.Request(url,
                                  headers={'Authorization': 'Basic %s' % auth})
-    response = urllib.request.urlopen(req)
+    if timeout is not None:
+        response = urllib.request.urlopen(req, timeout=timeout)
+    else:
+        response = urllib.request.urlopen(req)
     return json.loads(response.read().decode('utf-8'))
 
 
@@ -274,10 +277,10 @@ class ReplicationControl(object):
         t.do_run = True
         while t.do_run:
             run_queue = max(
-                _couchdb_request(
-                    self.source_url + '/_node/_local/_system')['run_queue'],
-                _couchdb_request(
-                    self.target_url + '/_node/_local/_system')['run_queue'])
+                _couchdb_request(self.source_url + '/_node/_local/_system',
+                                 timeout=60)['run_queue'],
+                _couchdb_request(self.target_url + '/_node/_local/_system',
+                                 timeout=60)['run_queue'])
             if run_queue > COUCHDB_MAX_RUN_QUEUE:
                 self.report_error()
             logging.debug('CouchDB run_queue: %d' % run_queue)

adrienverge commented on coucharchive:
```diff
@@ -74,7 +75,10 @@ def _couchdb_request(url):
+    if timeout is not None:
+        response = urllib.request.urlopen(req, timeout=timeout)
+    else:
+        response = urllib.request.urlopen(req)
```
You can drop this change: in case `timeout` is `None`, these 2 lines are equivalent :wink:

H--o-l replied:
I wasn't sure xD.
Can you tell me how you find that the 2 lines are equivalent in the None case?
In the meantime here is the pushed fix!
```diff
--- a/coucharchive
+++ b/coucharchive
@@ -75,10 +75,7 @@ def _couchdb_request(url, timeout=None):
     url = urlunsplit(parts)
     req = urllib.request.Request(url,
                                  headers={'Authorization': 'Basic %s' % auth})
-    if timeout is not None:
-        response = urllib.request.urlopen(req, timeout=timeout)
-    else:
-        response = urllib.request.urlopen(req)
+    response = urllib.request.urlopen(req, timeout=timeout)
     return json.loads(response.read().decode('utf-8'))

adrienverge replied:
Actually I was wrong: it would be the case if the default value was `None` here (which is very usual, but not here): https://github.com/python/cpython/blob/9af0e47b1705457bb6b327c197f2ec5737a1d8f6/Lib/urllib/request.py#L139

I'm suspecting that using `None` is still correct, because of https://github.com/python/cpython/blob/9af0e47b1705457bb6b327c197f2ec5737a1d8f6/Lib/urllib/request.py#L507, but can't be sure. So, sorry... what I said was wrong!

H--o-l replied:
Ok, no issue, thanks for the links!
I restore the first version.

adrienverge commented on coucharchive:
```diff
@@ -755,6 +818,9 @@ def main():
                         level=loglevel)
 
+    has_run_queue_control = \
+        args.action != 'load' and not args.no_run_queue_control
```
Hum OK why not!

H--o-l replied:
Without that the script crash.
Is it OK for you? Or you see a better solution?

adrienverge replied:
No, it's OK!

adrienverge reviewed (commented):
The PR is cool, the last point I have in mind is that 7 is a bit magical here: it doesn't have a real justification. Also, since it's not weigthed, getting a 8 will have the same effect as getting a 50.

To mitigate that, there are probably multiple solutions. Here is a simple one I'm thinking of:
- From my understanding of `ideal_number_of_replications()`, it seems that calling `report_error()` multiple times has more effect than calling it once (because the error will be weighted *n* times).
- So we could do that:
  ```diff
  -            if run_queue > COUCHDB_MAX_RUN_QUEUE:
  -                self.report_error()
  +            for _ in range(int(run_queue / COUCHDB_MAX_RUN_QUEUE)):
  +                self.report_error()
  ```
- Which means if `run_queue` is in `[0..6]` it's OK, if it's in `[7..13]` we report one error, if it's in `[14..20]` we report two errors, etc.

Could that work?

H--o-l commented:
Hey Adrien !

> The PR is cool, the last point I have in mind is that 7 is a bit magical here: it doesn't have a real justification. Also, since it's not weigthed, getting a 8 will have the same effect as getting a 50.
> 
> To mitigate that, there are probably multiple solutions. Here is a simple one I'm thinking of:
> 
> * From my understanding of `ideal_number_of_replications()`, it seems that calling `report_error()` multiple times has more effect than calling it once (because the error will be weighted _n_ times).
> * So we could do that:
>   ```diff
>   -            if run_queue > COUCHDB_MAX_RUN_QUEUE:
>   -                self.report_error()
>   +            for _ in range(int(run_queue / COUCHDB_MAX_RUN_QUEUE)):
>   +                self.report_error()
>   ```
> * Which means if `run_queue` is in `[0..6]` it's OK, if it's in `[7..13]` we report one error, if it's in `[14..20]` we report two errors, etc.
> 
> Could that work?

Cool idea, that will definitively works.
I just try locally on my computer and you can see in the trace below that the number of reported error is correctly incremented:
```
2020-01-20T10:21:53+0100 DEBUG: Currently running 110 replication workers
2020-01-20T10:21:53+0100 DEBUG: Ideal speed = fastest; current average speed = 5.0 rep/s
2020-01-20T10:21:53+0100 DEBUG: There were 79 CouchDB errors encountered in the last 5 minutes
2020-01-20T10:21:53+0100 DEBUG: CouchDB run_queue: 23
2020-01-20T10:21:54+0100 DEBUG: Currently running 109 replication workers
2020-01-20T10:21:54+0100 DEBUG: Ideal speed = fastest; current average speed = 5.0 rep/s
2020-01-20T10:21:54+0100 DEBUG: There were 82 CouchDB errors encountered in the last 5 minutes
```
I hope we will see a cool effect on a real cluster!

H--o-l commented:
Hey Adrien!

After many months this PRs doesn't seems to be a good idea, because it have a bad impact on performance, without, it seems, improving overall stability (CouchDB 500 error or HTTP timeout continue to happen).

Thus I will close it.
